### PR TITLE
Add get samples function to InstructionToSignals for JAX-jit usage

### PR DESCRIPTION
**Pull request by to24toro**
### Summary

This PR adds `get_samples ` to `InstructionToSignals` for JAX-jitting when using qiskit-pulse and removes the usage of `get_waveform` method of `SymbolicPulse`.

### Details and comments

`get_samples` function gets the envelope expression form` SymbolicPulse` and calls sympy.lambdify with numerical backend specified by Array class. The lambdified function is lru cached for performance.

### CODE REVIEW

**Comment by DanPuzzuoli** [comment]
Okay cool thanks.

> Feel free to on hold this until above PR is merged, or you can merge this PR without JIT test and later add the the test in a separate (follow up) PR.

I think it makes sense to hold on this PR, as I don't think the required functionality can truly be verified until terra is at a point where the `static_argnums` can be removed.

**Comment by nkanazawa1989** [comment]
Fair enough. Let's merge https://github.com/Qiskit/qiskit-terra/pull/9002 first.

File: test/dynamics/signals/test_pulse_to_signals.py
Comment (to24toro) [inline test/dynamics/signals/test_pulse_to_signals.py:271]: We expected this function will pass without using static_argnums after merging https://github.com/Qiskit/qiskit-terra/pull/9002

**Comment by DanPuzzuoli** [comment]
Hey @to24toro , there are two more tests I think it'd be good to add:
- Can you add a test case for computing gradients as well? You can take the same `jit` test, but change it so that at the end, instead of `jit`, you do `jax.jit(jax.grad(jit_func))(0.1)`. You'll have to change the the function so that it returns a real scalar (as `grad` only works on real scalars), so you could just change what the function returns to be `get_samples(instance).sum().real`. Testing this combination of gradding/jitting is important because it can introduce a bit more complexity into what JAX is doing, and sometimes raise errors where `jit`/`grad` don't raise them on their own.
- I'm also thinking we should add a test for `Solver` that compiles/differentiates a simulation of a pulse schedule with symbolic pulses. In the `test/dynamics/solvers/test_solver_classes.py` file there is a test class called `TestPulseSimulationJAX`. Can you add a test to this class that defines a function building a schedule with symbolic pulses, simulating it with a `Solver`, and returning the final state, and attempts to compile/differentiate it? This may be a little redundant with the other tests but it will be a good integration test.

One last point: the `TestJaxBase` class has two helper functions: `jit_wrap` and `jit_grad_wrap`, which can be used to jit a function, or to `jit(grad(func))` a function `func`, where the latter takes the sum and the real part of the output. These may eb helpful/convenient.

File: test/dynamics/signals/test_pulse_to_signals.py
Comment (to24toro) [inline test/dynamics/signals/test_pulse_to_signals.py:286]: I tried to use jit_wrap and jit_grad_wrap.
Very useful.

File: qiskit_dynamics/solvers/solver_classes.py
Comment (to24toro) [inline qiskit_dynamics/solvers/solver_classes.py:528]: This is related with the issue https://github.com/Qiskit/qiskit-dynamics/issues/175

**Comment by to24toro** [comment]
I modified two points at https://github.com/Qiskit/qiskit-dynamics/pull/149/commits/769e342a06f95d8232bd6475da3e7b46c80717d5 for JAX-jitting.

**Comment by DanPuzzuoli** [comment]
I've removed the "on hold" label, as you pointed out that terra has been sufficiently updated. Once the errors are resolved and I re-review we can merge this!

File: qiskit_dynamics/pulse/pulse_to_signals.py
Comment (DanPuzzuoli) [inline qiskit_dynamics/pulse/pulse_to_signals.py:127]: If `isinstance(inst.pulse, Waveform) == False`, is the only other option for `inst.pulse` to be a `SymbolicPulse`? Do we need to worry about any other cases in this code block?

File: qiskit_dynamics/pulse/pulse_to_signals.py
Comment (DanPuzzuoli) [inline qiskit_dynamics/pulse/pulse_to_signals.py:307]: Add type hint for `backend` as well as the return type.

File: qiskit_dynamics/pulse/pulse_to_signals.py
Comment (DanPuzzuoli) [inline qiskit_dynamics/pulse/pulse_to_signals.py:269]: Add type hint for return type.

File: test/dynamics/pulse/test_pulse_to_signals.py
Comment (DanPuzzuoli) [inline test/dynamics/pulse/test_pulse_to_signals.py:332]: Rewrite these tests to still call the `InstructionToSignals` converter. `get_samples` and `lru_cache_expr` are both internal functions, so ideally we won't directly test them. There are exceptions to this in Dynamics, but in this case I don't think it's necessary. The inputs/outputs of these functions are directly fed from/to the converter, so I don't think anything is gained by directly testing these functions, as opposed to verifying the behaviour of the converter on a symbolic pulse.

Maybe change the name of this to `test_SymbolicPulse`.

File: test/dynamics/pulse/test_pulse_to_signals.py
Comment (DanPuzzuoli) [inline test/dynamics/pulse/test_pulse_to_signals.py:350]: Similar to the preceding comment, I don't think it's necessary to have a separate class for checking the behaviour of `get_samples`. These tests could be moved to the previous class, by rewriting them to verify that the converter returns the same thing, whether you pass in a `Waveform` or a `SymbolicPulse`. 

The `jit` test is great and very important, but again, it can be moved to the previous class and rewritten to work with the converter.

File: test/dynamics/solvers/test_solver_classes.py
Comment (DanPuzzuoli) [inline test/dynamics/solvers/test_solver_classes.py:1267]: Good test here. Can you make the schedule more complicated by adding `ShiftPhase`, `SetPhase`, `ShiftFrequency`, and `SetFrequency` instructions? I believe you updated the PR so that these could also be jitted through, but that is not being tested here.

Can you also add a derivative test here as well, not just `jit`.

**Review by DanPuzzuoli — CHANGES_REQUESTED** [review]
Aside from current test failing issue (which seems to just be a sympy issue), I have some comments about reorganizing the new tests a little.

Looks good though.

File: test/dynamics/solvers/test_solver_classes.py
Comment (to24toro) [reply test/dynamics/solvers/test_solver_classes.py:1267]: add the new test with https://github.com/Qiskit/qiskit-dynamics/pull/149/commits/3f214c574e2f611b635dabde4dc21e150a411697

File: test/dynamics/pulse/test_pulse_to_signals.py
Comment (to24toro) [reply test/dynamics/pulse/test_pulse_to_signals.py]: improved and changed the name at https://github.com/Qiskit/qiskit-dynamics/pull/149/commits/e2c5148dade2da23c125e1471580e32046669d45

File: qiskit_dynamics/pulse/pulse_to_signals.py
Comment (to24toro) [reply qiskit_dynamics/pulse/pulse_to_signals.py]: added at https://github.com/Qiskit/qiskit-dynamics/pull/149/commits/4dab802b8e400282a022df72341da5aff353b5b0

File: qiskit_dynamics/pulse/pulse_to_signals.py
Comment (to24toro) [reply qiskit_dynamics/pulse/pulse_to_signals.py]: added at https://github.com/Qiskit/qiskit-dynamics/pull/149/commits/4dab802b8e400282a022df72341da5aff353b5b0

**Comment by to24toro** [comment]
My author and commit name are not correct. So I modified them and force-pushed.

File: qiskit_dynamics/pulse/pulse_to_signals.py
Comment (DanPuzzuoli) [inline qiskit_dynamics/pulse/pulse_to_signals.py:49]: I think something's happened here - this is undoing some changes I had made in a previous PR (that I don't think you are intentionally trying to change).

File: qiskit_dynamics/pulse/pulse_to_signals.py
Comment (DanPuzzuoli) [inline qiskit_dynamics/pulse/pulse_to_signals.py:71]: These two code blocks are similarly changing some documentation I had changed previously. Maybe some issue with a previous attempt at merging `main` into this branch?

File: qiskit_dynamics/pulse/pulse_to_signals.py
Comment (DanPuzzuoli) [inline qiskit_dynamics/pulse/pulse_to_signals.py:137]: Similarly here, this is undoing some changes I had made.

File: test/dynamics/pulse/test_pulse_to_signals.py
Comment (DanPuzzuoli) [reply test/dynamics/pulse/test_pulse_to_signals.py:341]: For this I mean to fully get rid of this test class, and to move the cases it's testing into the preceding test class. by calling `InstructionToSignals` directly.

File: test/dynamics/pulse/test_pulse_to_signals.py
Comment (DanPuzzuoli) [inline test/dynamics/pulse/test_pulse_to_signals.py:373]: E.g. this test should use `InstructionToSignals` instead of `get_samples` directly.

File: test/dynamics/pulse/test_pulse_to_signals.py
Comment (DanPuzzuoli) [inline test/dynamics/pulse/test_pulse_to_signals.py:380]: This is a great test, that could be moved to the preceding class.

File: test/dynamics/pulse/test_pulse_to_signals.py
Comment (to24toro) [reply test/dynamics/pulse/test_pulse_to_signals.py:341]: The class you said moving to is `TestPulseToSignals`?
If I move to the class, we fail to pass the test because `TestPulseToSignals` is not a subclass of `TestJaxBase`.
On the other hand, if `TestPulseToSignals` inherits `TestJaxBase`, all the test in `TestPulseToSignals` will be skipped in python test and executed in only JAX test.

File: qiskit_dynamics/pulse/pulse_to_signals.py
Comment (to24toro) [reply qiskit_dynamics/pulse/pulse_to_signals.py]: I am so sorry to confuse you. As you say, it is not my intention.  I restored to that of main at .

File: test/dynamics/pulse/test_pulse_to_signals.py
Comment (DanPuzzuoli) [reply test/dynamics/pulse/test_pulse_to_signals.py:341]: I understand now. Maybe just rename this class then to `TestPulseToSignalsJAXTransformations` or something along these lines.

Comment (to24toro) [reply test/dynamics/pulse/test_pulse_to_signals.py]: rename and fix at https://github.com/Qiskit/qiskit-dynamics/pull/149/commits/454f1194463df8cfe5bf002343e81fa92fdc04b4 and https://github.com/Qiskit/qiskit-dynamics/pull/149/commits/2814cf0587ceabc6354cbf628560536bcad24577

File: qiskit_dynamics/pulse/pulse_to_signals.py
Comment (DanPuzzuoli) [inline qiskit_dynamics/pulse/pulse_to_signals.py:53]: My last question is: did you intentionally remove these spaces? Deleting these causes this whole block of text to be a single paragraph in the docs. I prefer these as separate paragraphs, as each is highlighting something different. Unless there is a specific reason for this, can you put the spaces back in.

Comment (to24toro) [reply qiskit_dynamics/pulse/pulse_to_signals.py]: Sorry. I didn't take into account the paragraph change. I returned back at https://github.com/Qiskit/qiskit-dynamics/pull/149/commits/c3338c5b6ddd56a51cdd8fc156eff25174d6e2be.

**Review by DanPuzzuoli — APPROVED** [review]

**Pull request merged**
